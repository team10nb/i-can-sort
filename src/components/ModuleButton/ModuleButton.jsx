/*
    Author: Shiliang Chen, Yuting Jiang
*/

import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import ButtonBase from "@material-ui/core/ButtonBase";
import Typography from "@material-ui/core/Typography";
import "@fontsource/roboto";

export default function ModuleButton(props) {
    const {width, height, color, preOne, image, onClick} = props;
    const borderWidth = preOne ? "3px" : "0px";
    const presentHeight = preOne ? height - 3 : height;

    // styles of this module button
    const useStyles = makeStyles((theme) => ({
        root: {
            // boxShadow: "2px 8px 30px 2px #888888",
            borderTopLeftRadius: 17,
            borderTopRightRadius: 17,
            borderBottomRightRadius: 0,
            borderBottomLeftRadius: 0,
            borderColor: color,
            borderStyle: "solid",
            borderWidth: borderWidth,
            borderBottomWidth: "0px",
            marginTop: 5,
            marginBottom: 0,
            display: "flex",
            flexWrap: "wrap",
            minWidth: 200,
            minHeight: 180,
            width: width,
            height: presentHeight,
        },
        image: {
            position: "relative",
            height: presentHeight,
            width: "100%",
            // [theme.breakpoints.down("xs")]: {
            //     width: "100% !important", // Overrides inline-style
            //     height: 100,
            // },
            "&:hover, &$focusVisible": {
                zIndex: 1,
                "& $imageBackdrop": {
                    backgroundImage: `url(${image.gif})`,
                    backgroundColor: "",
                    opacity: 1,
                },
                "& $imageTitle": {
                    border: "4px solid currentColor",
                    opacity: 0,
                },
            },
        },
        focusVisible: {},
        imageButton: {
            borderTopLeftRadius: 17,
            borderTopRightRadius: 17,
            position: "absolute",
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            color: theme.palette.common.white,
        },
        imageSrc: {
            borderTopLeftRadius: 17,
            borderTopRightRadius: 17,
            // backgroundImage: `url(${props.image.static})`,
            position: "absolute",
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            backgroundSize: "cover",
            backgroundPosition: "center 40%",
        },
        imageBackdrop: {
            borderTopLeftRadius: 15,
            borderTopRightRadius: 15,
            backgroundImage: `url(${props.image.static})`,
            backgroundSize: "100% 100%",
            backgroundRepeat: "no-repeat",
            backgroundPosition: "center 40%",
            position: "absolute",
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            // backgroundColor: theme.palette.common.grey,
            backgroundColor: "#d5d5d5",
            opacity: 1,
            transition: theme.transitions.create("opacity"),
        },
        imageTitle: {
            position: "absolute",
            right: 0,
            bottom: 0,
            // padding: `${theme.spacing(2)}px ${theme.spacing(4)}px ${
            //     theme.spacing(1) + 6
            // }px`,
            paddingBottom: 5,
            paddingRight: 10,
            // fontFamily: "Roboto",
            fontFamily: "Quicksand",
            fontWeight: 1000,
            fontSize: 20,
            color: "white",
        },
    }));

    const classes = useStyles();

    return (
        <div className={classes.root}>
            <ButtonBase
                focusRipple
                key={image.title}
                className={classes.image}
                focusVisibleClassName={classes.focusVisible}
                onClick={onClick}
            >
                
                {/* background of the button */}
                <span
                    className={classes.imageSrc}
                    style={{
                        // backgroundImage: `url(${props.image.static})`,
                        backgroundColor: "#d5d5d5",
                    }}
                />

                {/* a layer beyond the background */}
                <span className={classes.imageBackdrop} />

                {/* title */}
                <span className={classes.imageButton}>
                    <Typography
                        component='span'
                        variant='subtitle1'
                        color= 'inherit'
                        className={classes.imageTitle}
                    >
                        {image.title}
                    </Typography>
                </span>

            </ButtonBase>
        </div>
    );
}
