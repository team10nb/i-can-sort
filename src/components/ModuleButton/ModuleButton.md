### **Developer**

Shiliang Chen, Yuting Jiang

Reference:

MATERIAL-UI (https://material-ui.com)

fontsource

### **Introduction**

Used in conjunction with a Module component as a button for user to enter a specific module of an sorting algorithm.

This button contains an image and a gif to indicate the sorting process. Gif will be shown when the cursor hovers.

### **Import**

```html
import { makeStyles } from "@material-ui/core/styles"; import ButtonBase from
"@material-ui/core/ButtonBase"; import Typography from
"@material-ui/core/Typography"; import "@fontsource/roboto";
```

### **Props**

| Name    | Type   | Default | Description                                                                   |
| ------- | ------ | ------- | ----------------------------------------------------------------------------- |
| width   | number |         | The length of the button.                                                     |
| height  | number |         | The height of the button.                                                     |
| color   | string | #3f51b5 | The color of the border.                                                      |
| preOne  | bool   | false   | If true, increase the width by 6px, indicating this is the last accessed one. |
| image   | object |         | The image and gif to be displayed on the button.                              |
| onClick | fn     |         | The function to be triggered when the button is clicked.                      |

### **Demo**

```jsx

const image = {
        static: "./logo512.png",
        gif: "catch.gif",
        title: "Bubble sort",
        width: "20%",
};
<ModuleButton width={200} height={200} color={'#1565c0'} preOne = {false} image={image}/>
```
