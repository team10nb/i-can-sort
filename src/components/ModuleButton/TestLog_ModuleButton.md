reason/description里面写为什么test fail了
对**组件代码**做修改和修改的理由写在每个代码块前的注释里（由于没有通过test，而对**原码**做的修改）
有必要的话，贴一下test的代码
第一个version可以不写注释 最后一个version(pass的version)可以不写reason

#### ComponentName: ModuleButton

|Test No.|Date|Pass/Fail|Failed test(s)|Reason/Description|Modified version|
|-------|----|---------|-------------------|----------------|----------------|
|1 | 1.22 | Fail | All | No code has been constructed |version 1|
|2 | 1.22 | Pass |/|Initial version is created|version 2|

##### Code - version 2
```javascript
import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import ButtonBase from "@material-ui/core/ButtonBase";
import Typography from "@material-ui/core/Typography";
import "@fontsource/roboto"

export default function ModuleButton(props) {
    const useStyles = makeStyles((theme) => ({
        root: {
            boxShadow: "2px 8px 30px 2px #888888",
            borderTopLeftRadius: 10,
            borderTopRightRadius: 10,
            borderBottomRightRadius: 2,
            borderBottomLeftRadius: 2,
            marginTop: 5,
            marginBottom: 5,
            display: "flex",
            flexWrap: "wrap",
            minWidth: 200,
            minHeight: 200,
            width: props.width,
            height: props.height,
        },
        image: {
            position: "relative",
            height: 200,
            width: "100%",
            // [theme.breakpoints.down("xs")]: {
            //     width: "100% !important", // Overrides inline-style
            //     height: 100,
            // },
            "&:hover, &$focusVisible": {
                zIndex: 1,
                "& $imageBackdrop": {
                    backgroundImage: `url(${props.image.gif})`,
                    backgroundColor: "",
                    opacity: 1,
                },
                "& $imageTitle": {
                    border: "4px solid currentColor",
                    opacity: 0,
                },
            },
        },
        focusVisible: {},
        imageButton: {
            borderTopLeftRadius: 10,
            borderTopRightRadius: 10,
            position: "absolute",
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            display: "flex",
            alignItems: "center",
            justifyContent: "center",
            color: theme.palette.common.white,
        },
        imageSrc: {
            borderTopLeftRadius: 10,
            borderTopRightRadius: 10,
            backgroundImage: `url(${props.image.static})`,
            position: "absolute",
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            backgroundSize: "cover",
            backgroundPosition: "center 40%",
        },
        imageBackdrop: {
            borderTopLeftRadius: 10,
            borderTopRightRadius: 10,
            backgroundImage: `url(${props.image.static})`,
            backgroundSize: "100% 100%",
            backgroundRepeat: "no-repeat",
            backgroundPosition: "center 40%",
            position: "absolute",
            left: 0,
            right: 0,
            top: 0,
            bottom: 0,
            backgroundColor: theme.palette.common.grey,
            opacity: 0.4,
            transition: theme.transitions.create("opacity"),
        },
        imageTitle: {
            position: "absolute",
            padding: `${theme.spacing(2)}px ${theme.spacing(4)}px ${
                theme.spacing(1) + 6
            }px`,
            fontFamily: "Roboto",
            fontSize: 20,
        },
    }));

    const classes = useStyles();

    return (
        <div className={classes.root}>
            <ButtonBase
                // focusRipple
                key={props.image.title}
                className={classes.image}
                focusVisibleClassName={classes.focusVisible}
                onClick={props.onClick}
            >
                <span
                    className={classes.imageSrc}
                    style={{
                        // backgroundImage: `url(${props.image.static})`,
                        backgroundColor: 'grey',
                    }}
                />
                <span className={classes.imageBackdrop} />

                <span className={classes.imageButton}>
                    <Typography
                        component='span'
                        variant='subtitle1'
                        color='inherit'
                        className={classes.imageTitle}
                    >
                        {props.image.title}
                    </Typography>
                </span>
            </ButtonBase>
        </div>
    );
}

```


