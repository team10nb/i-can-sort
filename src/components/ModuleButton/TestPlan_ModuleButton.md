## Test Plan
--------
<br>

#### Component: ModuleButton

|Test Name|Description|Operation|Expected Outcome|
|----|------|----------------|------------|
|should contain bubblesort|This component should be able to display title properly|try find the demo text|demo text found|
| should call onclick | should be able to response to click event and call the onclick function | Mock click event. | onClick function will be called once. |
| should match snapshot | Compare new snapshot with old one. |  | Snapshots are the same. |

