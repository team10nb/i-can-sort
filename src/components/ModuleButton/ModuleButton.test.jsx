/*
    Author: Shiliang Chen, Yuting Jiang
*/

import { render } from '@testing-library/react';
import userEvent from "@testing-library/user-event";
import ModuleButton from './ModuleButton';

test('should contain Bubble Sort', () => {
    
    const mockHandleClick = jest.fn();
    
    const image = {
        static: "./logo512.png",
        gif: "sorting.gif",
        title: "Bubble sort",
    };

    const props = {
        image: image,
        width: 200,
        height: 200,
        onClick: mockHandleClick,
    };

    const {getByText} = render(<ModuleButton {...props} />);
    getByText("Bubble sort");
    

})

test('should call onClick', () => {
    
    const mockHandleClick = jest.fn();
    
    const image = {
        static: "./logo512.png",
        gif: "sorting.gif",
        title: "Bubble sort",
    };

    const props = {
        image: image,
        width: 200,
        height: 200,
        onClick: mockHandleClick,
    };

    const {getByText} = render(<ModuleButton {...props} />);
    const button = getByText("Bubble sort");
    userEvent.click(button);
    expect(mockHandleClick).toBeCalledTimes(1);

})


test('should match Snapshot', () => {
    
    const mockHandleClick = jest.fn();
    
    const image = {
        static: "./logo512.png",
        gif: "sorting.gif",
        title: "Bubble sort",
    };

    const props = {
        image: image,
        width: 200,
        height: 200,
        onClick: mockHandleClick,
    };

    const content = render(<ModuleButton {...props} />);
    expect(content).toMatchSnapshot();

})

