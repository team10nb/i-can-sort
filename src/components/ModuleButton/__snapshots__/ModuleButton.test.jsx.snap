// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-15"
      >
        <button
          class="MuiButtonBase-root makeStyles-image-16"
          tabindex="0"
          type="button"
        >
          <span
            class="makeStyles-imageSrc-19"
            style="background-color: grey;"
          />
          <span
            class="makeStyles-imageBackdrop-20"
          />
          <span
            class="makeStyles-imageButton-18"
          >
            <span
              class="MuiTypography-root makeStyles-imageTitle-21 MuiTypography-subtitle1 MuiTypography-colorInherit"
            >
              Bubble sort
            </span>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-15"
    >
      <button
        class="MuiButtonBase-root makeStyles-image-16"
        tabindex="0"
        type="button"
      >
        <span
          class="makeStyles-imageSrc-19"
          style="background-color: grey;"
        />
        <span
          class="makeStyles-imageBackdrop-20"
        />
        <span
          class="makeStyles-imageButton-18"
        >
          <span
            class="MuiTypography-root makeStyles-imageTitle-21 MuiTypography-subtitle1 MuiTypography-colorInherit"
          >
            Bubble sort
          </span>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
